### PR TITLE
Fixes PHP error in `lib/block-supports/align.php` & other undefined $block_type->supports instances

### LINKED ISSUE #24922: Gutenberg 8.8. contains the PHP notice: Undefined property: WP_Block_Type::$supports
**Describe the bug**

After update from 8.7.0 to 8.8.0 appears errors on edit post screen:

```
Notice: Undefined property: WP_Block_Type::$supports in ...\gutenberg\lib\block-supports\align.php on line 14
Notice: Undefined property: WP_Block_Type::$supports in ...\gutenberg\lib\block-supports\colors.php on line 14
Notice: Undefined property: WP_Block_Type::$supports in ...\gutenberg\lib\block-supports\typography.php on line 14
Notice: Undefined property: WP_Block_Type::$supports in ...\gutenberg\lib\block-supports\custom-classname.php on line 14
Notice: Undefined property: WP_Block_Type::$supports in ...\gutenberg\lib\block-supports\generated-classname.php on line 14
```

This bug was already in version 8.6.0 #24149 and fixed in 8.7.0.
Problem returned after PR #24351

**To reproduce**
1. Enable WP_DEBUG
2. Go to edit post screen
4. See errors

**Editor version (please complete the following information):**
- WordPress version: 5.5
- gutenberg plugin 8.8.0

### PR DESCRIPTION
**Pull request by aristath**
closes #24922 

Working on the master branch, I got an error log full of these messages just now:
```
PHP Notice:  Undefined property: WP_Block_Type::$supports in wp-content/plugins/gutenberg/lib/block-supports/align.php on line 14
PHP Stack trace:
PHP   1. {main}() index.php:0
PHP   2. require() index.php:17
PHP   3. require_once() wp-blog-header.php:13
PHP   4. require_once() wp-load.php:37
PHP   5. require_once() wp-config.php:92
PHP   6. do_action() wp-settings.php:540
PHP   7. WP_Hook->do_action() wp-includes/plugin.php:478
PHP   8. WP_Hook->apply_filters() wp-includes/class-wp-hook.php:311
PHP   9. gutenberg_register_block_supports() wp-includes/class-wp-hook.php:287
PHP  10. gutenberg_register_alignment_support() wp-content/plugins/gutenberg/lib/block-supports/index.php:17
```

This PR adds an extra check to make sure the property exists before using it.

### CODE REVIEW

**Comment by youknowriad** [comment]
This is not the only place where $block->supports is used, it's also used in all the other block-supports hooks. Maybe we should instead add an empty array there if it's not defined when registering a block instead of having to add this check everywhere.

**Comment by aristath** [comment]
Hmmm looks like a fix for this one was pushed in WP-Core in 5.5.0, and the `WP_Block_Type` now has a `supports` property as it should.
I haven't been able to reproduce this issue, so perhaps I was just using WP5.4 at the time.

Closing this PR as it's no longer relevant.

**Comment by youknowriad** [comment]
Apparently the issue is still present even on 5.5 see #24922 we might want to revisit this PR and apply it to all block supports.

**Comment by aristath** [comment]
@youknowriad I reopened the PR, grep'd Gutenberg's codebase for `->supports` and pushed an additional commit to add extra checks where needed.

File: lib/block-supports/generated-classname.php [hunk 1]
```diff
@@ -41,7 +41,10 @@ function gutenberg_get_block_default_classname( $block_name ) {
  * @return array Block CSS classes and inline styles.
  */
 function gutenberg_apply_generated_classname_support( $attributes, $block_attributes, $block_type ) {
-	$has_generated_classname_support = gutenberg_experimental_get( $block_type->supports, array( 'className' ), true );
+	$has_generated_classname_support = false;
```
Comment (youknowriad) [inline lib/block-supports/generated-classname.php:44]: the default value is true here.

File: lib/block-supports/custom-classname.php [hunk 2]
```diff
@@ -35,7 +38,10 @@ function gutenberg_register_custom_classname_support( $block_type ) {
  * @return array Block CSS classes and inline styles.
  */
 function gutenberg_apply_custom_classname_support( $attributes, $block_attributes, $block_type ) {
-	$has_custom_classname_support = gutenberg_experimental_get( $block_type->supports, array( 'customClassName' ), true );
+	$has_custom_classname_support = false;
```
Comment (youknowriad) [inline lib/block-supports/custom-classname.php:41]: The default value is true here.

File: lib/block-supports/custom-classname.php [hunk 3]
```diff
@@ -11,7 +11,10 @@
  * @param WP_Block_Type $block_type Block Type.
  */
 function gutenberg_register_custom_classname_support( $block_type ) {
-	$has_custom_classname_support = gutenberg_experimental_get( $block_type->supports, array( 'customClassName' ), true );
+	$has_custom_classname_support = false;
```
Comment (youknowriad) [inline lib/block-supports/custom-classname.php:14]: The default value is true here.

File: lib/block-supports/generated-classname.php (hunk 1 above)
Comment (aristath) [reply lib/block-supports/generated-classname.php]: OMG you're absolutely right... fixed :+1:

File: lib/block-supports/custom-classname.php (hunk 2 above)
Comment (aristath) [reply lib/block-supports/custom-classname.php]: fixed

File: lib/block-supports/custom-classname.php (hunk 3 above)
Comment (aristath) [reply lib/block-supports/custom-classname.php]: fixed

**Review by youknowriad — APPROVED** [review]
Not sure how to test because I don't really reproduce but code-wise it looks good.

**Comment by aristath** [comment]
> Not sure how to test because I don't really reproduce

The only way I've been able to reproduce was by using WP5.4 with the latest Gutenberg plugin... Or simply replacing the `wp-includes/class-wp-block-type.php` with a previous version from https://github.com/WordPress/WordPress/blob/5.4/wp-includes/class-wp-block-type.php (though that may throw additional and irrelevant errors).

**Comment by youknowriad** [comment]
you could argue that it's a good fix regardless since Gutenberg the plugin does support WP 5.4 too.

**Comment by skorasaurus** [comment]
> > Not sure how to test because I don't really reproduce
> 
> The only way I've been able to reproduce was by using WP5.4 with the latest Gutenberg plugin... Or simply replacing the `wp-includes/class-wp-block-type.php` file with a previous version from https://github.com/WordPress/WordPress/blob/5.4/wp-includes/class-wp-block-type.php (though that may throw additional and irrelevant errors).

As another data point, I have the same exact errors as https://github.com/WordPress/gutenberg/issues/24922 and for me, I only was only able to reproduce it on Gutenberg 8.8.0 and WordPress 5.4.2.... Once I upgraded core to 5.5, error went away.

**Pull request merged**